feature/matcher/src/main/java/com/simprints/matcher/usecases/MatchResultSet.kt

```diff
@@ -1,40 +1,48 @@
 package com.simprints.matcher.usecases
 
 import com.simprints.matcher.MatchResultItem
-import java.util.TreeSet
+import java.util.concurrent.ConcurrentSkipListSet
+import java.util.concurrent.atomic.AtomicReference
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.concurrent.withLock
 
 internal class MatchResultSet<T : MatchResultItem>(
     private val maxSize: Int = MAX_RESULTS,
 ) {
-    private var lowestConfidence: Float = 0f
+    private val lowestConfidence = AtomicReference(0f)
+    private val lock = ReentrantLock()
 
-    private val treeSet = TreeSet(
+    private val skipListSet = ConcurrentSkipListSet(
         compareByDescending<T> { it.confidence }.thenByDescending { it.subjectId },
     )
 
     fun add(element: T): MatchResultSet<T> {
-        if (lowestConfidence > element.confidence) {
-            // skip adding if the last element is greater than the current element
+        // Use a lock to ensure thread safety during the entire add operation
+        lock.withLock {
+            // Only perform this optimization when we know the set is at max capacity
+            if (skipListSet.size >= maxSize && lowestConfidence.get() > element.confidence) {
```

Comment:
Are atomic reference for "lowestConfidence" and concurrent collection required if the whole block is synchronised?

Reply:
I'm tempted to dismiss this concern for structures with max 10 results. However, if you feel strongly about this, we can do a quick benchmark and verify TreeSet indeed works fine in real concurrent situations (CoSync reading with N threads)!?